// openelectricity/dist/node/index.js
"use strict";
/**
 * OpenElectricity API Client
 * Main entry point for the OpenElectricity API client library
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.stripTimezone = exports.makeAware = exports.isAware = exports.getNetworkTimezoneOffset = exports.getNetworkTimezone = exports.getLastCompleteInterval = exports.DataTable = exports.OpenElectricityClient = exports.default = void 0;
// Export the client
__exportStar(require("./client"), exports);
var client_1 = require("./client");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return client_1.OpenElectricityClient; } });
Object.defineProperty(exports, "OpenElectricityClient", { enumerable: true, get: function () { return client_1.OpenElectricityClient; } });
// Export DataTable class
var datatable_1 = require("./datatable");
Object.defineProperty(exports, "DataTable", { enumerable: true, get: function () { return datatable_1.DataTable; } });
// Export datetime utilities
var datetime_1 = require("./datetime");
Object.defineProperty(exports, "getLastCompleteInterval", { enumerable: true, get: function () { return datetime_1.getLastCompleteInterval; } });
Object.defineProperty(exports, "getNetworkTimezone", { enumerable: true, get: function () { return datetime_1.getNetworkTimezone; } });
Object.defineProperty(exports, "getNetworkTimezoneOffset", { enumerable: true, get: function () { return datetime_1.getNetworkTimezoneOffset; } });
Object.defineProperty(exports, "isAware", { enumerable: true, get: function () { return datetime_1.isAware; } });
Object.defineProperty(exports, "makeAware", { enumerable: true, get: function () { return datetime_1.makeAware; } });
Object.defineProperty(exports, "stripTimezone", { enumerable: true, get: function () { return datetime_1.stripTimezone; } });
